### PR TITLE
BB-184 Update worker formula and add more Gunicorn worker configuration

### PR DESCRIPTION
**Pull request by xirdneh**
Configuration Pull Request
---

This PR updates the `workers` formula to improve readability. The formula used to read `(cpus - 1) * mul + mul` where `cpus` is the number of cpus and `mul` is a multiplier which value is set in the yaml variables. If we expand this formula we can simplify it to `cpus * mul`:
```
(cpus - 1) * mul + mul
(mul * cpus) - mul + mul
mul * cpus
```
This PR also adds three more gunicorn config variables:
[threads](http://docs.gunicorn.org/en/stable/settings.html#threads) : This will allow us to use multiple threads for each worker. Making it possible to run more workers and keeping gunicorn memory footprint low.
[worker_class](http://docs.gunicorn.org/en/stable/settings.html#worker-class) : If `threads` are used then it is possible to specify which type of thread class gunicorn will use. By default it will use `gthread` if any number of `threads` if specified.
[max_requests_jitter](http://docs.gunicorn.org/en/stable/settings.html#max-requests-jitter) : This will allow us to specify a max number of requests before gunicorn restart a worker. By doing this we help handle any memory leak happening in the application.

Reasoning:

We have experimented with a different amount of workers as well as max requests before gunicorn restart the workers. Using a number around 14 of workers have helped us better use all the resources in different cases. This is why we suggest adding gunicorn's `thread` and `worker_class` that way other users can run more workers while keeping gunicorn's memory footprint almost the same. Using 14 `sync` workers ( the default type of workers without threads) also works perfectly, hence using `threads` and `worker_class` is completely optional.
We've also added the `max_requests_jitter` configuration so users can specify how many requests will each worker handle before being restarted by gunicorn. This has allowed us to avoid memory leak or high memory consumption by one or multiple workers in different cases. We usually use `max_request` but in the latest gunicorn they've introduced `max_request_jitter` which will make gunicorn use a random number of max request between 0 and this value. This avoids having all threads restart at the same time and make the application sluggish.

Make sure that the following steps are done before merging:

  - [ ] A DevOps team member has approved the PR if it is code shared across multiple services and you don't own all of the services.
  - [x] Are you adding any new default values that need to be overridden when this change goes live? If so:
    - [ ] Update the appropriate internal repo (be sure to update for all our environments)
    - [ ] If you are updating a secure value rather than an internal one, file a DEVOPS ticket with details.
    - [ ] Add an entry to the CHANGELOG.
  - [ ] If you are making a complicated change, have you performed the proper testing specified on the [Ops Ansible Testing Checklist](https://openedx.atlassian.net/wiki/display/EdxOps/Ops+Ansible+Testing+Checklist)?  Adding a new variable does not require the full list (although testing on a sandbox is a great idea to ensure it links with your downstream code changes).
  - [ ] Think about how this change will affect Open edX operators.  Have you updated the wiki page for the next Open edX release?

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @xirdneh! I've created [OSPR-3654](https://openedx.atlassian.net/browse/OSPR-3654) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

**Comment by feanil** [comment]
@xirdneh it looks like you've got some conflicts, can you fix those and let me know when this is ready to be reviewed.

**Comment by xirdneh** [comment]
@feanil This is ready for review. Thanks!

**Comment by feanil** [comment]
Sounds good, I'll queue it up for the devops team at edX to review.

File: playbooks/roles/edxapp/defaults/main.yml [hunk 1]
```diff
@@ -588,7 +588,19 @@ EDXAPP_CODE_JAIL_LIMITS:
...
+#   cms: <num worker threads>
 EDXAPP_WORKERS: !!null
+EDXAPP_WORKER_THREADS: !!null
+EDXAPP_WORKER_CLASS: !!null
+# Gunicorn will restart each worker after x amount of
+# requests processed. The x value will be randomly selected between
+# 1 and this value (default 1500)
+EDXAPP_CMS_MAX_REQUEST_JITTER: 1500
+EDXAPP_LMS_MAX_REQUEST_JITTER: 1500
+EDXAPP_ANALYTICS_DATA_TOKEN: ""
```
Comment (jdmulloy) [inline playbooks/roles/edxapp/defaults/main.yml]: This doesn't seem relevant to this PR. Is this an accidental inclusion?
```suggestion
```

File: playbooks/roles/edxapp/defaults/main.yml [hunk 2]
```diff
@@ -588,7 +588,19 @@ EDXAPP_CODE_JAIL_LIMITS:
...
+EDXAPP_ANALYTICS_DATA_URL: ""
```
Comment (jdmulloy) [inline playbooks/roles/edxapp/defaults/main.yml:602]: This doesn't seem relevant to this PR. Is this an accidental inclusion?
```suggestion
```

Comment (xirdneh) [reply playbooks/roles/edxapp/defaults/main.yml:602]: @jdmulloy Yes, sorry. I'll remove it. Thanks.

Comment (xirdneh) [reply playbooks/roles/edxapp/defaults/main.yml:602]: @jdmulloy I've remove the variable, mind taking a look? Thanks!

**Comment by kaizoku** [comment]
Thanks @xirdneh, looks good on our test instances for this :+1:

- [✓] I tested this: (examined configuration on test instances after configuration)
- [✓] I read through the code
- [n/a] ~I checked for accessibility issues~
- [n/a] ~Includes documentation~

**Comment by xirdneh** [comment]
@jdmulloy This has been updated, can you take a look?

**Comment by xirdneh** [comment]
@jdmulloy @feanil Any update on this? thanks!

**Comment by feanil** [comment]
I've queued it up for jmulloy to review again.

FYI @natabene

**Comment by viadanna** [comment]
@natabene @jdmulloy I'm taking this over from Josue and we'll be reevaluating the approach. This PR can be closed and we'll open another one in the future.

**Comment by openedx-webhooks** [comment]
@xirdneh Even though your pull request wasn’t merged, please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/configuration&entry.867055334=https://github.com/edx/configuration/pull/5185&entry.1484655318=https://github.com/xirdneh&entry.752974735=2019-06-12+17:14&entry.1917517419=2019-08-29+15:26&entry.2133058324=No) so we can improve your experience in the future.
